File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    static boolean[] visited;
    static int min = Integer.MAX_VALUE;
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        visited = new boolean[words.length];
        if(!Arrays.asList(words).contains(target)) {
            return 0;
        }
        dfs(begin, target, words, 0);
        
        return min == Integer.MAX_VALUE ? 0 : min;
    }
    static void dfs(String begin, String target, String[] words, int answer) {
        
        if(begin.equals(target)){
            min = Math.min(min, answer);
            return;
        }
        
        for(int i = 0; i < words.length; i++){
            
            int cnt = 0;
            for(int j = 0; j < begin.length(); j++){
                if(begin.charAt(j) == words[i].charAt(j)){
                    cnt++;
                }
            }
            
            if(cnt == begin.length() - 1 && !visited[i]){
                visited[i] = true;
                dfs(words[i], target, words, answer + 1);
                visited[i] = false;
            }
        }
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

bool isOneDiff(string a, string b){
    int diff = 0;
    for(int i = 0; i < a.length(); i++){
        if(a[i] != b[i]){
            diff++;
        }
    }
    return diff == 1;
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    bool hasTarget = false;
    for(string w: words){
        if(w == target){
            hasTarget = true;
            break;
        }
    }
    
    if(!hasTarget){
        return 0;
    }
    
    queue<pair<string, int>> q;
    vector<bool> visited(words.size(), false);
    
    q.push({begin, 0});
    
    while(!q.empty()){
        string cur = q.front().first;
        int cnt = q.front().second;
        q.pop();
        
        if(cur == target){
            return cnt;
        }
        
        for(int i = 0; i < words.size(); i++){
            if(!visited[i] && isOneDiff(cur, words[i])){
                visited[i] = true;
                q.push({words[i], cnt + 1});
            }
        }
    }
    return 0;
}